python/tk_multi_publish2/api/item.py

```diff
@@ -0,0 +1,1012 @@
+            return get_parent_pixmap()
+        else:
+            if default_pixmap_path:
+                # return default
+                return QtGui.QPixmap(default_pixmap_path)
+            else:
+                return None
+
+    @icon.setter
+    def icon(self, pixmap):
+        """Sets the icon."""
+        self._icon_pixmap = pixmap
+
+    @property
+    def is_root(self):
+        """
+        Returns ``True`` if this the root :ref:`publish-api-item` in the tree,
+        ``False`` otherwise.
+        """
+        return self.parent is None
+
+    @property
+    def local_properties(self):
+        """
+        A :class:`~.api.PublishData` instance that houses item properties local
+        to the current :class:`~.base_hooks.PublishPlugin` instance. As such, it
+        is expected that this property is only accessed from within a publish
+        plugin. Attempts to access this property outside of a publish plugin
+        will raise an ``AttributeError``.
+
+        This property behaves like the local storage in python's threading
+        module, except here, the data is local to the current publish plugin.
+
+        You can get and set values for this property using standard dictionary
+        notation or via dot notation.
+
+        It is important to consider when to set a value via
+        :py:attr:`~properties`` and when to use :py:attr:`~local_properties`.
+
+        Setting the values on :py:attr:`~properties` is a way to globally share
+        information between publish plugins. Values set via
+        :py:attr:`~local_properties` will only be applied during the execution
+        of the current plugin (similar to python's ``threading.local`` storage).
+
+        A common scenario to consider is when you have multiple publish plugins
+        acting on the same item. You may, for example, want the ``publish_name``
+        and ``publish_version`` properties to be shared by each plugin, while
+        setting the remaining properties on each plugin instance since they will
+        be specific to that plugin's output. Example:
+
+        .. code-block:: python
+
+            # set shared properties on the item (potentially in the collector or
+            # the first publish plugin). these values will be available to all
+            # publish plugins attached to the item.
+            item.properties.publish_name = "Gorilla"
+            item.properties.publish_version = "0003"
+
+            # set specific properties in subclasses of the base file publish
+            # (this class). first publish plugin...
+            item.local_properties.publish_template = "asset_fbx_template"
+            item.local_properties.publish_type = "FBX File"
+
+            # in another publish plugin...
+            item.local_properties.publish_template = "asset_abc_template"
+            item.local_properties.publish_type = "Alembic Cache"
+
+        .. note:: If you plan to serialize your publish tree, you may run into
```

Comment:
Should we link to Python docs on how to implement custom serialization logic for complex objects?

Reply:
Now that I've removed comments about pickling, since we're not doing that but dump everything to json instead, I've updated the comment to clarify that everything should be JSON serializable.

If using JSON serializable objects is not always feasible and this becomes an issue we'll expose a serialize/unserialize methods on the plugins.